internal/controller/reinstall/reinstall.go

```diff
@@ -372,3 +379,279 @@ func (r *ReinstallHandler) finalizeReinstallRequest(
+	}
+
+	for _, secret := range clusterInstanceSecrets {
+		err := applyPreservationLabel(secret)
+		if err != nil {
+			return reconcile.Result{},
+				fmt.Errorf("encountered an error applying preservation label to ClusterInstance Secret, err: %w", err)
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+// ensureDataIsPreserved preserves resources
+func (r *ReinstallHandler) ensureDataIsPreserved(
+	ctx context.Context,
+	log *zap.Logger,
+	clusterInstance *v1alpha1.ClusterInstance,
+) (result reconcile.Result, err error) {
+
+	result = reconcile.Result{}
+	err = nil
+
+	var (
+		preservationDataBackedupCondition, clusterIdentityDetectedCondition *metav1.Condition
+	)
+
+	defer func() {
+		patch := client.MergeFrom(clusterInstance.DeepCopy())
+		updateRequired := false
+
+		if preservationDataBackedupCondition != nil {
+			updateRequired = setReinstallStatusCondition(clusterInstance, *preservationDataBackedupCondition)
+		}
+
+		if clusterIdentityDetectedCondition != nil {
+			if changed := setReinstallStatusCondition(clusterInstance, *clusterIdentityDetectedCondition); changed {
+				updateRequired = true
+			}
+		}
+
+		if updateRequired {
+			result.Requeue = true
+			updateErr := conditions.PatchCIStatus(ctx, r.Client, clusterInstance, patch)
+			if updateErr != nil {
+				result.RequeueAfter = requeueWithShortInterval
+				err = logAndWrapUpdateFailure(log, err, updateErr,
+					"failed to update reinstall preservation condition(s)")
+			}
+		}
+	}()
+
+	if clusterInstance.Spec.Reinstall.PreservationMode == v1alpha1.PreservationModeNone {
+		preservationDataBackedupCondition = reinstallPreservationDataBackedupConditionStatus(
+			metav1.ConditionFalse, v1alpha1.PreservationNotRequired, "PreservationMode is set to None.")
+		clusterIdentityDetectedCondition = reinstallClusterIdentityDataDetectedConditionStatus(
+			metav1.ConditionFalse, v1alpha1.PreservationNotRequired, "PreservationMode is set to None.")
+		return
+	}
+
+	preservationDataBackedupCondition = findReinstallStatusCondition(clusterInstance,
+		v1alpha1.ReinstallPreservationDataBackedup)
+	clusterIdentityDetectedCondition = findReinstallStatusCondition(clusterInstance,
+		v1alpha1.ReinstallClusterIdentityDataDetected)
+
+	setInProgress := false
+	if preservationDataBackedupCondition == nil ||
+		preservationDataBackedupCondition.Reason == string(v1alpha1.Initialized) {
+		preservationDataBackedupCondition = reinstallPreservationDataBackedupConditionStatus(
+			metav1.ConditionFalse, v1alpha1.InProgress, "Data is being preserved")
+		setInProgress = true
+	}
+
+	if clusterIdentityDetectedCondition == nil ||
+		clusterIdentityDetectedCondition.Reason == string(v1alpha1.Initialized) {
+		clusterIdentityDetectedCondition = reinstallClusterIdentityDataDetectedConditionStatus(
+			metav1.ConditionFalse, v1alpha1.InProgress, "Data is being preserved")
+		setInProgress = true
+	}
+
+	if setInProgress {
+		result.Requeue = true
```

Comment:
Why requeue here?

Reply:
To update the ClusterInstance condition status to `InProgress`.